### ISSUE TITLE
What about bgfx?

### ISSUE THREAD

**Issue by woollybah** [davecamp/BMax-Shader-Framework#1]
Wouldn't it be far better to sit this framework on top of something like [bgfx](https://github.com/bkaradzic/bgfx), which basically handles all the cross-platform/graphics rendering stuff using a standard shader language.
It has support for GL, ES, DirectX, Metal, and Vulkan, apparently.
It also does batching, which can often improve the pipeline efficiency.

I had started work on implementing this library previously - I have ported examples and whatnot - but since I'm crap with shaders in general (and really bad at asking people for help), I didn't get it to a point it would be generally useful - where we might have a set of generic shaders for use with Max2D, for example.

The idea of off-loading all the messy work of implementing this stuff is that you let someone else (and all their users) test the core stuff, and you then just need to worry about the BlitzMax level integration and support...

If you are interested in going down this direction, I can dig out my previous work on it and get it up-to-date with the latest code in the library?

**Comment by davecamp** [comment]
That sounds interesting to me 👍 

> If you are interested in going down this direction, I can dig out my previous work on it and get it up-to-date with the latest code in the library?

Yeah sure that'd be great. Don't spend too much time updating it at the mo though as BMax2D is quite basic ( <= version 1.3 gl shaders by default, <= dx shader level 3 for d3d9 and <= shader level 5 for d3d11, most people will want just pixel shaders at first too ) and what you have already may be enough?

**Comment by GWRon** [comment]
> [..] but since I'm crap with shaders in general (and really bad at asking people for help)

Yeah, but there is Col - our man in charge when it comes to GFX stuff!

My main concern is, that bgfx might be a "too big beast" for what Col is willed to tackle. It seems to play nice with SDL and GLFW but _I_ cannot imagine what would be needed to incorporate it.

I assume it would completely replace max2D then... bgfx sounds mighty - and batching stuff is surely _pretty_ nice.

**Comment by davecamp** [comment]
Thinking about it, yes... I would think it does replace Max2D completely?

**Comment by woollybah** [comment]
Also, since Apple has now deprecated OpenGL/ES on the forthcoming OS's, it would be nice if we had the option of a different render layer on Mac/iOS without having to rewrite everything - again.
And as we know, 32-bit is already deprecated on macOS...

yay...

**Comment by GWRon** [comment]
@davecamp 
So what's your plan - do you want to tinker about bgfx + "max2dNG.mod"? If so it might be a good idea to collaboratively "layout" the module - and maybe use the chance to ask people in the syntaxbomb.com forum with a certain "shader knowledge" (like GaborD) - or "agk experience" (some more users). So we can find out what would a more current approach "need" and what not.

I think it would need some simple support to adjust the way things are "projected" (means standard is a 2D projection while it should be easy to insert/hook-in your own stuff to add 3D stuff). Same for the render pipeline.
Default should be the "sequential" approach (what was added first is drawn first) as this corresponds to the current max2D-approach. But maybe it would be good to have manually createable "batches" 

something like this:
```BlitzMax
StartRenderBatch() 'returns a batch-object you could refer later on - eg as StartBatch(useBatch)
DrawImage(img1, 0, 0)
DrawImage(img2,100, 100)
EndRenderBatch()
```
`StartRenderBatch(useBatch:TMax2DNGRenderBatch=null)` is only a wrapper to the original function in the new `TMax2DNG"-class. You pass an existing batch to "append", or skip passing one to create a new one (which is returned for external access).
Think we had a similar approach for the "Render2Texture"-functionality.

**Comment by davecamp** [comment]
> So what's your plan 

I was going to continue on with this shader framework. When Brucey releases what code he has so far for bgfx ( @woollybah  It doesn't matter of the state of the code of what you have done already, we can all join minds and finish it together) then I would take a look at what's been done and see where we can take it.

GaborD does have good experience with using shaders too so I'm sure his help would be very valuable ( he also seems keen to play with full shader support in BMax ) to help get that balance between user code difficulty/exposure versus whats not worth exposing.

Brucey... over to you my friend :) when do you think we can get our greasy hands on your existing bgfx code?

**Comment by GWRon** [comment]
@woollybah (@"brucey" is someone else :-))
We talked about bgfx in connection with pxjs in april 2015 ... so only 3 years ago :-). And nope, I never received a bgfx.mod to try out (which is why I asked for content in the bgfx.mod-repo in 2016). Time flies by.

**Comment by davecamp** [comment]
@ the Brucey / woollybah thing...
I edited it. Would a message be sent out to the owner of the Brucey handle? hehe... oops :p

**Comment by woollybah** [comment]
Am I aiming for this to work with both legacy BlitzMax and NG?
I imagine Derron would say yes to that. But remember there's more limited potential mileage for 32-bit Mac and Linux variants as time goes on...

For integration as a "PUB" type low level module, and for use as something for a Graphics driver to plug into - where driver creates the window, and we create the graphics context/renderer, I'm going to have to come at things from a different angle than my original code.

I've recently rewritten the SDL graphics driver to work better with SDLWindows, so I've mostly got my head around that level of integration now.
There are currently 3 "parts" that make up the base stuff - bgfx, bx and bimg. bx has common stuff. bimg is an imaging thing. bgfx is the main engine. There's also some tools for compiling/converting shaders and textures.

Leave it with me, and I'll see what I can come up with.

**Comment by davecamp** [comment]
> Leave it with me, and I'll see what I can come up with.

> and really bad at asking people for help

You don't have to do it alone... where and how can I help :-)

**Comment by GWRon** [comment]
Of course a vanilla support would be nice (others might stay with vanilla too..for their old projects, so the new bgfx base would just make it work on other platforms / newer DX etc.). Also in regards to bug-free-ness I trust in vanilla for now ...albeit NG matured a huge bit over the time already.

I am btw for a NG.mod base...no need to use pub.mod or brl.mod. Especially if it is not compatible to vanilla.

If it eases the pain to use NG (interfaces etc...) then go with that!

**Comment by woollybah** [comment]
Latest progress is here : https://github.com/bmx-ng/gfx.mod
There's a stub Max2D implementation which does enough to get a window open, but the core stuff of actually drawing things is yet to be done.
It may be that we can port the GL2 versions of Max2D which is already set up for rendering to shaders. Or maybe it's better (i.e. more efficient) to rethink how we are drawing stuff?
If we can take advantage of bgfx's pipeline optimisations that would be cool.

There's stuff about bgfx shaders here : https://bkaradzic.github.io/bgfx/tools.html#shader-compiler-shaderc. I'll see about supplying shaderc binaries (I don't see any need for a game to be compiling shaders at runtime, so I wasn't planning on integrating it further).

**Comment by GWRon** [comment]
@davecamp 
Did you find any time to play with gfx.mod + bgfx yet?

Think it is the best stuff for days with >30°C :-)

**Comment by davecamp** [comment]
I have been playing with it in c++ to familiarise myself with how it works, but not with BMax just yet. It's a good engine in the sense that it does well to minimise state changes via command buffer stateless rendering. This may well mean to not use the existing Max2D interface to take advantage of its features.

> Think it is the best stuff for days with >30°C :-)

Hehe I'm an outdoors person myself, coding comes second for me :p

**Comment by GWRon** [comment]
So ... summer has gone and winter is yet to come. Any thoughts on reviving that stuff?

Why am I asking? Just thought about adding a suggestion about `StartRenderBatch()` and `FinishRenderBatch()` ... and saw that I already posted such a thing in this issue here ;)

I would like to append something. A `TRenderBatch` could be a graphic object like the current "canvas". so they all have convenience helpers like: `AddShader(shader)` (or maybe also "Pre" and "Post"-Shader or at whatever stages shaders would come in handy). So at the end you could set a shader to a whole canvas.
Canvases can be: the complete graphics window, a texture (eg Render2Texture... so maybe not a texture but some kind of canvas existing during r2t creation) - or it could be a `TRenderBatch`-instance (which creates a canvas too).

Essence would be the _optional_ creation of "to render" graphics objects which can get processed by some shaders.

@woollybah 
Do you still think "bgfx.mod" is worth to get implemented? I think it is useful if there is someone out there utilizing it - and I am not sure if @davecamp has time and is motivated to do so?

Thoughts?
